db/write_batch.cc

```diff
@@ -2612,11 +2609,116 @@ Status WriteBatchInternal::InsertInto(
   return s;
 }
 
+// This class computes checksums for a WriteBatch.
+// This is WIP -- gonzo does not really fully know (yet) what he is doing.
+class ChecksumUpdater : public WriteBatch::Handler {
+ public:
+  explicit ChecksumUpdater(WriteBatch::ProtectionInfo* prot_info)
+      : prot_info_(prot_info) {}
+
+  ~ChecksumUpdater() override {}
+
+  Status PutCF(uint32_t cf, const Slice& key, const Slice& val) override {
+    return update_checksum(cf, key, val);
+  }
+
+  Status DeleteCF(uint32_t /* cf */, const Slice& /* key */) override {
+    // TODO gonzo: does it make sense to checksum the key (no value)?
+    return do_nothing();
+  }
+
+  Status SingleDeleteCF(uint32_t /* cf */, const Slice& /* key */) override {
+    // TODO gonzo: does it make sense to checksum the key (no value)?
+    return do_nothing();
+  }
+
+  Status DeleteRangeCF(uint32_t /* cf */, const Slice& /* begin_key */,
+                       const Slice& /* end_key */) override {
+    // TODO gonzo: does it make sense to checksum the keys (no values)?
+    return do_nothing();
+  }
+
+  Status MergeCF(uint32_t cf, const Slice& key, const Slice& val) override {
+    return update_checksum(cf, key, val);
+  }
+
+  Status PutBlobIndexCF(uint32_t cf, const Slice& key,
+                        const Slice& val) override {
+    return update_checksum(cf, key, val);
+  }
+
+  Status MarkBeginPrepare(bool /* unprepare */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
+
+  Status MarkEndPrepare(const Slice& /* xid */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
+
+  Status MarkCommit(const Slice& /* xid */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
+
+  Status MarkCommitWithTimestamp(const Slice& /* xid */,
+                                 const Slice& /* ts */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
+
+  Status MarkRollback(const Slice& /* xid */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
+
+  Status MarkNoop(bool /* empty_batch */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
+
+ private:
+  Status update_checksum(uint32_t cf, const Slice& key, const Slice& val) {
+    if (key.size() > size_t{port::kMaxUint32}) {
+      return Status::InvalidArgument("key is too large");
+    }
+    if (val.size() > size_t{port::kMaxUint32}) {
+      return Status::InvalidArgument("value is too large");
+    }
+    if (prot_info_) {
+      prot_info_->entries_.emplace_back(
+          ProtectionInfo64().ProtectKVO(key, val, kTypeValue).ProtectC(cf));
```

Comment:
We need to use the optype corresponding to the operation, not always kTypeValue.

Reply:
Of course, silly me. Done as suggested.